lock_api/src/mutex.rs

```diff
@@ -458,17 +460,21 @@ impl<'a, R: RawMutex + 'a, T: ?Sized + 'a> MutexGuard<'a, R, T> {
     /// used as `MutexGuard::try_map(...)`. A method would interfere with methods of
     /// the same name on the contents of the locked data.
     #[inline]
-    pub fn try_map<U: ?Sized, F>(s: Self, f: F) -> Result<MappedMutexGuard<'a, R, U>, Self>
+    pub fn try_map<U: 'a, F>(s: Self, f: F) -> Result<OwnedMappedMutexGuard<'a, R, U>, Self>
     where
-        F: FnOnce(&mut T) -> Option<&mut U>,
+        F: FnOnce(&'a mut T) -> Option<U>,
     {
+        let s = mem::ManuallyDrop::new(s);
+
         let raw = &s.mutex.raw;
-        let data = match f(unsafe { &mut *s.mutex.data.get() }) {
+        let data = unsafe { &mut *s.mutex.data.get() };
+
+        let data = match f(data) {
```

Comment:
This will leak the mutex guard if `f(data)` panics.

Reply:
I'm trying to work out what the correct behaviour here would actually be - I've tried to work out a way to cause an invalid reference using the `f` closure but as far as I see it the lifetime bounds make it fine. I think just using `mem::forget(s)` after the `match` instead of using `mem::ManuallyDrop` is the correct behaviour.

I would have preferred to use `catch_unwind`/`resume_unwind` to make the control flow more explicit but unfortunately that's not available with `no_std` right now.